daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,217 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.dsom.walker
+
+import java.io.File
+
+import org.apache.daffodil.compiler.Compiler
+import org.apache.daffodil.dsom.{Choice, ComplexTypeBase, ElementBase, GroupRef, ModelGroup, Root, Sequence, SimpleTypeBase, Term}
+
+/**
+ * A class designed to walk the internal representation of a DFDL Schema File.
+ *
+ * There are numerous internal event handlers to be defined by the implementing class; each
+ * corresponds to "encountering" a particular element type in the DFDL Schema file.
+ *
+ * Though recursion is used here to define the walk, it is not advised to use recursion between
+ * these event handlers.  Instead, consider a stack-like structure, as the DFDL Schema structure
+ * as well as the recursive method call structure can be represented by trees.
+ * @param schemaFile an input DFDL Schema file to parse
+ * @tparam T the return type of walkDSOMSchema and onWalkEnd.  This should be a class used to
+ *           represent the structure of the Schema.  You could always make it Unit if you don't
+ *           want these methods to return anything.
+ */
+abstract class AbstractDSOMWalker[T](schemaFile: File) {
+  /**
+   * The root element of the DFDL Schema.  This will be the starting point for the traversal
+   */
+  private final val schemaSetRoot: Root = Compiler().compileFile(schemaFile).sset.root
```

Comment:
I like Steve's suggestion to separate compilation and walking, and provide a DSOM walk method which takes a Listener class (if I understand correctly).  I have worked on someone else's Visitor class for an ANTLR v4 grammar and then written my own Listener class for the same ANTLR v4 grammar.  In my experience, the Listener class for the same use case was smaller and more maintainable since ANTLR took care of visiting every part of the parse tree (smaller chance for mistakes) and I was able to implement only a handful of listener methods that I actually needed.  ANTLR listeners can't return values directly from method calls, but I was able to attach these return values to specific parts of the parse tree (by modifying the ANTLR grammar to provide fields for these values) and have higher up listener methods collect these values from lower down parts of the parse tree.  Collecting return values may become harder if you can't store them in the tree as you walk the tree like I did, though.

In case people aren't familiar with ANTLR visitors and walkers, someone has blogged about them (<https://saumitra.me/blog/antlr4-visitor-vs-listener-pattern/>) and said:

> Good thing about listener pattern is that its all automatic and even if you don’t write any parse tree walker, ANTLR will figure out and trigger the enter and exit method for each rule. This is a huge benefit for translation type of usecases.

Reply:
This all makes a lot more sense idiomatically.  In the latest commit I opted for adding this method in ProcessorFactory:
`def walkDSOM[T](walker: AbstractDSOMWalker[T]): T = walker.walkFromRoot(sset.root)`
AbstractDSOMWalker now also does not take in any constructor parameters.